mne/preprocessing/maxwell.py

```diff
@@ -41,6 +41,88 @@
 
 
+@verbose
+def maxwell_filter_prepare_emptyroom(
+    raw_er, *, raw, bads='from_raw', verbose=None
+):
+    """Prepare an empty-room recording for Maxwell filtering.
+
+    Empty-room data by default lacks certain properties that are required to
+    ensure running :func:`~mne.preprocessing.maxwell_filter` will process the
+    empty-room recording the same way as the experimental data. This function
+    preconditions empty-room raw data instance accordingly so it can be used
+    for Maxwell filtering.
+
+    Specifically, this function will:
+
+    * Compile the list of bad channels according to the ``bads`` parameter.
+    * Inject the device-to-head transformation matrix from the experimental
+      recording into the empty-room recording.
+    * Set the same montage on the empty-room recording as the experimental
+      recording.
+
+    Parameters
+    ----------
+    raw_er : instance of Raw
+        The empty-room recording. It will not be modified.
+    raw : instance of Raw
+        The experimental recording, typically this will be the reference run
+        used for Maxwell filtering.
+    bads : 'from_raw' | 'union' | list of str
+        How to populate the list of bad channel names to be injected into
+        the empty-room recording. If ``'from_raw'`` (default) the list of bad
+        channels will be overwritten with that of ``raw``. If ``'union'``, will
+        use the union of bad channels in ``raw`` and ``raw_er``. Note that
+        this may lead to additional bad channels in the empty-room in
+        comparison to the experimental recording. If a list, you can explicitly
+        provide the list of channel names to be marked as bad.
+
+        .. note::
+           Non-MEG channels are silently dropped from the list of bads.
+    %(verbose)s
+
+    Returns
+    -------
+    raw_er_prepared : instance of Raw
+        A copy of the passed empty-room recording, ready for Maxwell filtering.
+
+    Notes
+    -----
+    .. versionadded:: 1.1
+    """
+    _validate_type(item=raw_er, types=BaseRaw, item_name='raw_er')
+    _validate_type(item=raw, types=BaseRaw, item_name='raw')
+    _validate_type(item=bads, types=(list, str), item_name='bads')
+    if isinstance(bads, str):
+        _check_option(
+            parameter='bads', value=bads, allowed_values=['from_raw', 'union']
+        )
+
+    raw_er_prepared = raw_er.copy()
+    del raw_er  # just to be sure
+
+    # handle bads; only keep MEG channels
+
+    if bads == 'from_raw':
+        bads = raw.info['bads']
+    elif bads == 'union':
+        bads = sorted(
+            set(raw.info['bads'] + raw_er_prepared.info['bads'])
+        )
+    bads = [ch_name for ch_name in bads
+            if ch_name.startswith('MEG')]
+    raw_er_prepared.info['bads'] = bads
+
+    # handle dev_head_t
+    raw_er_prepared.info['dev_head_t'] = raw.info['dev_head_t']
+
+    # handle montage
+    montage = raw.get_montage()
+    raw_er_prepared.set_montage(montage)
+
```

Comment:
You should add a comment somewhere that this is still incomplete because it does not set `raw.first_samp` to the same value

Reply:
> You should add a comment somewhere that this is still incomplete because it does not set `raw.first_samp` to the same value

I don't see why that would be needed..? (setting raw.first_samp, not the documentation)

Reply:
@larsoner said on Friday that it's necessary for continuous head motion compensation

Reply:
Does this mean we **should** alter `raw_er.first_samp`, or that we should simply document that we're **not** altering it..? But then, we'd need to advice the users what to do with that information..?

Reply:
I think you just need to do

`raw_er._first_samps = raw._first_samps`

is that correct @larsoner?

Reply:
No, not anymore. There is a `self._cropped_samp` now

https://github.com/mne-tools/mne-python/blob/dc88b7a18297818e7ab796d8e43755e499f324cc/mne/io/base.py#L579-L584

Also we should change the `meas_date` so that it matches as well, and then `set_annotations` using `raw.annotations`. (The `meas_date` change is necessary to make sure someone can `raw.set_annotations(...)` and `raw_erm.set_annotations(...)` after the fact and have it match, too.)

Reply:
Wait wait wait, what?

Why would anyone want to transfer annotations?

I don't want to touch the measurement data. This is what we rely in in MNE-BIDS to generate the correct session names for empty-room data. Also altering it kills provenance. 

If we ever add this it should not be the default behavior!

Reply:
> Why would anyone want to transfer annotations? It's empty-room data!

The guiding principle is that the file after this function should be as similar as possible to the actual data in terms of what happens to it in downstream functions (esp. maxwell_filter, and how it affects files).

One concrete example (that we account for in mnefun) is head movement. The `first_samp` needs to be correct. But also, the annotations should match, e.g., those produced and applied (before *or* after maxwell filtering!) by https://mne.tools/stable/generated/mne.preprocessing.annotate_movement.html. This knocks out sections of data, so you want to knock out the same time intervals in both the subject data and the empty room data.

The way to guarantee this is to 1) at the very least copy over annotations (in case this function was used before MF), and I would argue for consistency also 2) set `meas_date` (in case this function is used after MF).

> This is what we rely on in MNE-BIDS to generate the correct session names for empty-room data.

I would argue this is brittleness in MNE-BIDS, and we should perhaps find a workaround

> Also altering it kills provenance.

I don't fully understand this argument. We're intentionally overwriting properties (provenance I guess in some cases, including first samp?) of the empty room data, so it seems like an acceptable/expected effect.

Reply:
The way I understand it is the only reason you'd want to touch `meas_date` is to allow for setting annotations, not because it would otherwise make sense, right?

In this case, I'd be strongly in favor of changing the time in the annotations before applying them, and to leave the `meas_date` untouched.

I have to admit I still don't fully understand why one would want to do that in the first place – the empty-room recordings I've dealt with in the past were usually only a few minutes long, definitely much shorter than experimental runs. So we'd usually be unable to copy over all annotations anyway, simply because they'd be cropped on the right side of the time axis… 🤔

Reply:
> In this case, I'd be strongly in favor of changing the time in the annotations before applying them, and to leave the meas_date untouched.

This is more work for the user... exactly the kind of work we're trying to save them from having to do here. Instead of saying "just do raw_erm.set_annotations(annot)", we would have to say ... do some difficult / annoying time adjustment in annotations, then set them. And they will probably forget or not notice, as they would reasonably (I think) expect `raw_erm` to "look like" `raw` for the purposes of downstream MNE functions as much as possible.

> So we'd usually be unable to copy over all annotations anyway, simply because they'd be cropped on the right side of the time axis…

Yes you'll lose *some* if the durations don't match, but more consistency is better than less. And it can be especially important during fast head/unrealistic movements because these can lead to wacky head position estimates that shouldn't be used...

Reply:
One way to reduce risk of the `meas_date` / annot mismatch is to warn people that, if they want annotations to match (which they should if they're doing movecomp) they need to do `raw.set_annotations` *before* running the function. I could live with this even though I'd prefer setting `meas_date` to match. (We should also warn in the function that the `meas_date` will *not* match so annotations have to be dealt with differently.)

Reply:
> This is more work for the user... exactly the kind of work we're trying to save them from having to do here. Instead of saying "just do raw_erm.set_annotations(annot)", 

Errm no, this is what this function here would do for them …
Why would you expect them to do **this** step manually while we automate the rest?

Reply:
Because you'd want to set the annots **after** MW-filtering?

Reply:
> Because you'd want to set the annots after MW-filtering?

Yes